Modules/posixmodule.c

```diff
@@ -4800,43 +4802,70 @@ os__path_isdir_impl(PyObject *module, PyObject *path)
     }
 
     Py_BEGIN_ALLOW_THREADS
-    if (_path.fd != -1) {
-        hfile = _Py_get_osfhandle_noraise(_path.fd);
-        close_file = FALSE;
-    }
-    else {
-        hfile = CreateFileW(_path.wide, FILE_READ_ATTRIBUTES, 0, NULL,
-                            OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+    if (_path.wide) {    
+        if (_Py_GetFileInformationByName(_path.wide, FileStatBasicByNameInfo,
+                                         &statInfo, sizeof(statInfo))) {
+            if (!(statInfo.FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT)) {
+                slow_path = FALSE;
+                result = statInfo.FileAttributes & FILE_ATTRIBUTE_DIRECTORY;
+            } else if (!(statInfo.FileAttributes & FILE_ATTRIBUTE_DIRECTORY)) {
+                slow_path = FALSE;
+                result = 0;
+            }
+        } else {
+            switch(GetLastError()) {
+                case ERROR_FILE_NOT_FOUND:
+                case ERROR_PATH_NOT_FOUND:
+                case ERROR_NOT_READY:
+                case ERROR_BAD_NET_NAME:
```

Comment:
Four other errors can be excluded from the slow fallback path:

* `ERROR_BAD_NETPATH` - The network path was not found.
* `ERROR_BAD_PATHNAME` - The specified path is invalid.
* `ERROR_INVALID_NAME` - The filename, directory name, or volume label syntax is incorrect.
* `ERROR_FILENAME_EXCED_RANGE` - The filename or extension is too long.

```suggestion
                case ERROR_FILE_NOT_FOUND:
                case ERROR_PATH_NOT_FOUND:
                case ERROR_NOT_READY:
                case ERROR_BAD_NET_NAME:
                case ERROR_BAD_NETPATH:
                case ERROR_BAD_PATHNAME:
                case ERROR_INVALID_NAME:
                case ERROR_FILENAME_EXCED_RANGE:
```

Reply:
Thanks! I've gone ahead and added those cases for isdir, isfile, exists, & islink.